apps/s_client.c

```diff
@@ -798,6 +801,31 @@ static int new_session_cb(SSL *S, SSL_SESSION *sess)
     return 0;
 }
 
+static void keylog_callback(const SSL *ssl, const char *line)
+{
+    BIO *keylog = BIO_new_file(keylog_file, "a");
+    char buffer[256];
```

Comment:
Regarding ensuring each full line including the `\n` is written in one syscall, there *is* a BIO filter just for that.  `crypto/bio/bf_lbuf.c`.  With that one pushed on your base BIO, you can do all writes any way you want, they will be collected and only full lines will be written.